romancal/source_detection/source_detection_step.py

```diff
@@ -0,0 +1,232 @@
+        # calc_threshold_2d: choice of mean, median, or mode.
+        bkg_boxsize = integer(default=3)  # if calc_threshold_img or
+        # calc_threshold_2d: size of box in pixels for 2D background.
+        bkg_sigma = float(default=2.0) # if calc_threshold_img or
+        # calc_threshold_2d, n sigma for sigma clipping bkgrnd.
+        bkg_filter_size = integer(default=3) # if calc_threshold_img or
+        # calc_threshold_2d, size of gauss. kernel for background.
+        save_catalogs = boolean(default=False) # Save source catalog to file?
+        # Will overwrite an existing catalog of the same name.
+        output_cat_filetype = option('asdf', 'ecsv', default='asdf') # Used if
+        #save_catalogs=True - file type of output catalog.
+    """
+
+    def process(self, input):
+
+        with rdd.open(input) as input_model:
+
+            # remove units from data in this step.
+            # DAOStarFinder requires unitless input
+            if hasattr(input_model.data, "unit"):
+                self.data = input_model.data.value
+            else:
+                self.data = input_model.data
+
+            # mask DO_NOT_USE pixels?
+            if self.coverage_mask:
+                self.coverage_mask = (
+                    (dqflags.pixel["DO_NOT_USE"]) & input_model.dq
+                ).astype(bool)
+            else:
+                self.coverage_mask = None
+
+            # if a pre-determined threshold value for detection for the whole
+            # image is provided, use this
+            if self.scalar_threshold is not None:
+                threshold = float(self.scalar_threshold)
+                log.info(f"Using a detection threshold of {threshold}.")
+
+            # otherwise, if specified, calculate a scalar threshold from the
+            # image by calculating a 2D background image, using this to create
+            # a 2d threshold image, and using the median of the 2d threshold
+            # image as the scalar detection threshold for the whole image
+            elif self.calc_threshold is not None:
+                log.info("Determining detection threshold from image.")
+                bkg = self._calc_2D_background()
+                threshold_img = bkg.background + (
+                    self.snr_threshold * bkg.background_rms
+                )
+                threshold = np.median(threshold_img)
+                log.info(f"Calculated a detection threshold of {threshold} from image.")
+
+            log.info("Detecting sources with DAOFind, using entire image array.")
+            daofind = DAOStarFinder(
+                fwhm=self.kernel_fwhm,
+                threshold=threshold,
+                sharplo=self.sharplo,
+                sharphi=self.sharphi,
+                roundlo=self.roundlo,
+                roundhi=self.roundhi,
+                brightest=self.max_sources,
+                peakmax=self.peakmax,
+            )
+
+            sources = daofind(self.data, mask=self.coverage_mask)
```

Comment:
My recollection was that we found this behaved better if one subtracted the sky from the data rather than built it into the threshold?

Reply:
correct, fixed.